### PR TITLE
Async bug fixes

### PR DESCRIPTION
**Pull request by SergeyTeplyakov**
This PR contains fixes for all outstanding bugs related to async/iterator blocks pending for VS2015.

Some of them are actually affects VS2013 and previous versions as well.

Fixing this will unblocks us from releasing VS2015 support.

Would really appreciated for the review!

### CODE REVIEW

File: Foxtrot/Foxtrot/Extractor.cs [hunk 1]
```diff
@@ -2077,19 +2078,17 @@ public override void VisitAssignmentStatement(AssignmentStatement assignment)
           method.Contract.ContractInitializer = contractInitializerBlock;
           method.Contract.AsyncEnsures = asyncPostconditions;
 
+          
+          // Following replacement causes some weird issues for complex preconditions (like x != null && x.Length > 0)
+          // when CCRewriter is used with /publicsurface or Preconditions only.
+          // This fix could be temporal and proper fix would be applied in the future.
+          // After discussion this issue with original CC authors (Mike Barnett and Francesco Logozzo),
+          // we decided that this fix is safe and lack of Assume statement in the MoveNext method will not affect
+          // customers (neither CCRewriter customers now CCCheck customers).
```
Comment (krk) [inline Foxtrot/Foxtrot/Extractor.cs]: typo now -> nor

Comment (SergeyTeplyakov) [inline Foxtrot/Foxtrot/Extractor.cs]: Thanks. Will fix.

**Comment by mike-barnett** [comment]
Thanks for doing this! It looks great.

File: Foxtrot/Tests/TestInputs.xml [hunk 2]
```diff
@@ -370,7 +380,7 @@
      Compiler="CS"
      />
   <TestFile
-     Name="Foxtrot\Tests\Sources\Async12.cs"
+     Name="Foxtrot\Tests\Sources\Async_EnsuresWithCapturedForAll.cs"
```
Comment (sharwell) [inline Foxtrot/Tests/TestInputs.xml]: :question: Did you mean to remove `Async12` here?

Comment (SergeyTeplyakov) [inline Foxtrot/Tests/TestInputs.xml]: Reverted it back.

**Pull request merged**